mkosi

```diff
@@ -86,7 +87,25 @@ def run(cmdline: List[str], execvp: bool = False, **kwargs: Any) -> subprocess.C
         assert not kwargs
         os.execvp(cmdline[0], cmdline)
     else:
-        return subprocess.run(cmdline, **kwargs)
+        # CTRL+C is sent to the entire process group. We delay its handling in mkosi itself so the subprocess
+        # can exit cleanly before doing mkosi's cleanup. If we don't do this, we get device or resource is
+        # busy errors when unmounting stuff later on during cleanup.
```

Comment:
humn.. but why we don't want to really kill the subprocesses and wait for completion?  It sucks if it's a long running process or even if the process gets stuck for some reason.

I think we will need to be smarter than this to really kill any process an guarantee no one is left behind. @poettering  ideas?

Reply:
The subprocess still receives the SIGINT signal and (if it's well-behaved) should exit shortly after receiving it, this just makes mkosi wait until the subprocess exits before continuing. If we don't, we get race conditions since if the subprocess is still running, unmounting stuff later on in the mkosi cleanup will fail.

systemd does something similar by waiting (by default) for 30 seconds after sending SIGTERM before killing a process.

What I could add is an override so that if CTRL+C is pressed again while we're waiting for a subprocess to exit after a first CTRL+C, we stop waiting in mkosi and proceed with cleanup regardless of whether the subprocess is still running or not.

Reply:
> What I could add is an override so that if CTRL+C is pressed again while we're waiting for a subprocess to exit after a first CTRL+C, we stop waiting in mkosi and proceed with cleanup regardless of whether the subprocess is still running or not.

then we would need to be sure to kill the sub-processes that are still using a mountpoint. I thought we had something like that already

Reply:
Currently, we don't. I'm not sure if anything more complex than this can be implemented without moving from the `subprocess.run` interface to `Popen` since `subprocess.run` blocks until the child process exits.